src/common.speech/ServiceRecognizerBase.ts

```diff
@@ -404,119 +344,105 @@ export abstract class ServiceRecognizerBase implements IDisposable {
         }
     }
 
-    protected receiveMessageOverride: () => any = undefined;
+    protected receiveMessageOverride: () => Promise<void> = undefined;
 
-    protected receiveMessage = (): Promise<IConnection> => {
-        return this.fetchConnection().on((connection: IConnection): Promise<IConnection> => {
-            return connection.read()
-                .onSuccessContinueWithPromise((message: ConnectionMessage) => {
+    protected async receiveMessage(): Promise<void> {
+        try {
+            if (this.privIsDisposed) {
+                // We're done.
+                return;
+            }
 
-                    if (this.receiveMessageOverride !== undefined) {
-                        return this.receiveMessageOverride();
-                    }
-                    if (this.privIsDisposed) {
-                        // We're done.
-                        return PromiseHelper.fromResult(undefined);
-                    }
+            let connection = await this.fetchConnection();
```

Comment:
Not for this PR. In `receiveMessage()`, should we start a new connection? The `fetchConnection()` will try to open a new connection if the current connection is closed (mostly due to an error). But the re-opened connection will return no message as we never send any request in this connection.

Reply:
It's arguable. If audio is flowing (or should be, maybe it's suspended in the wait throttle) one or the other will open the connection. If message flow was stopping, I expect the queue drain from the old connection to have stopped it.


---
In reply to: [456227814](https://github.com/microsoft/cognitive-services-speech-sdk-js/pull/198#discussion_r456227814) [](ancestors = 456227814)